File: cpp/PetShop.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

class PetShop{
  private:
    // Atribut petshop
    // menggunakan list untuk bisa menyimpan banyak data dan mudah jika ingin menambahkan data baru
    list<int> id;
    list<string> name;
    list<string> category;
    list<int> price;

  public:

    // Constructor
    // saat objek di instansiasi, maka data akan langsung diisi oleh data default
    PetShop(){
      // Data default
      id.push_back(1);
      name.push_back("Kucing Persia");
      category.push_back("Hewan");
      price.push_back(98999);

      id.push_back(2);
      name.push_back("Penyu Madura");
      category.push_back("Hewan");
      price.push_back(215999);

      id.push_back(3);
      name.push_back("Katak Acumalaka");
      category.push_back("Hewan");
      price.push_back(14199);
    }

    // SETTER
    // setter untuk menambahkan data seluruh data baru ke list (tidak bisa set satu persatu attribute)
    void setPetShop(int id, string name, string category, int price){
      // Menambahkan data baru ke list menggunakan pushback untuk menambahkan di akhir list
      this->id.push_back(id);
      this->name.push_back(name);
      this->category.push_back(category);
      this->price.push_back(price);
    }

    // GETTER
    // getter ini akan mereturn seluruh data yang ada
    list<int> getId(){
      // Mengembalikan data id
      return this->id;
    }
    list<string> getName(){
      // Mengembalikan data name
      return this->name;
    }
    list<string> getCategory(){
      // Mengembalikan data category
      return this->category;
    }
    list<int> getPrice(){
      // Mengembalikan data price
      return this->price;
    }


    // Method
    // menampilkan data dalam bentuk tabel yang rapi
    // menggunakan library <iomanip> untuk mempermudah dalam membuat tabel
    void showPetShop() {
      // Inisialisasi iterator
      // digunakan untuk mengakses data dalam list produk
      auto id_it = id.begin();
      auto name_it = name.begin();
      auto category_it = category.begin();
      auto price_it = price.begin();

      // Header tabel
      cout << "\n🐾 Daftar Hewan di Pet Shop 🐾\n";
      cout << left  // Membuat teks rata kiri
             << setw(5) << "ID"        // Lebar kolom ID: 5 karakter
             << setw(20) << "Name"     // Lebar kolom Name: 20 karakter
             << setw(15) << "Category" // Lebar kolom Category: 15 karakter
             << setw(10) << "Price"    // Lebar kolom Price: 10 karakter
             << endl;

      // Garis pemisah antara header tabel dan baris data
      cout << string(50, '-') << endl;

      // Menampilkan data dalam bentuk tabel
      while (id_it != id.end()) {
        cout << left  // membuat teks rata kiri agar rapi
             << setw(5) << *id_it          // ID
             << setw(20) << *name_it       // Nama Hewan
             << setw(15) << *category_it   // Kategori Hewan
             << "Rp " << fixed << *price_it // Harga dengan format tetap
             << endl;

        // Menggeser iterator ke elemen berikutnya
        ++id_it;
        ++name_it;
        ++category_it;
        ++price_it;
      }
      cout << "\n";
  }

    // Mencari hewan berdasarkan nama dan menampilkan data lengkap jika ditemukan
    bool findName(string searchName) {
      // Inisialisasi iterator
      // digunakan untuk mengakses data dalam list produk
      auto id_it = id.begin();
      auto name_it = name.begin();
      auto category_it = category.begin();
      auto price_it = price.begin();

      cout << "\n🔍 Mencari '" << searchName << "'...\n";
      // algoritma yang dicari hanyalah looping sampai akhir
      // jika data ditemukan, maka akan menampilkan data lengkapnya
      while (name_it != name.end()) {
          if (*name_it == searchName) {
              // Menampilkan data lengkap dengan format tabel
              cout << "\n✅ Data ditemukan!\n";
              // header tabel
              cout << left << setw(5) << "ID"
                   << setw(20) << "Name"     
                   << setw(15) << "Category" 
                   << setw(10) << "Price" << endl; 
              cout << string(50, '-') << endl;
              // data yang ditemukan
              cout << left << setw(5) << *id_it
                   << setw(20) << *name_it
                   << setw(15) << *category_it
                   << "Rp " << fixed << *price_it << endl;
              return true;
          }
          // iterator untuk next elemen
          ++id_it;
          ++name_it;
          ++category_it;
          ++price_it;
      }
      // jika data tidak ditemukan, maka akan menampilkan pesan
      cout << "❌ Data tidak ditemukan.\n";
      return false;
    }

    // menghapus data berdasarkan nama yang akan dicari
    void deletePetShop(string name){
      // Inisialisasi iterator
      // digunakan untuk mengakses data dalam list produk
      auto id_it = this->id.begin();
      auto name_it = this->name.begin();
      auto category_it = this->category.begin();
      auto price_it = this->price.begin();

      cout << "\n🔍 Menghapus '" << name << "'...\n";
      // mencari dulu data nya, jika ditemukan maka hapus data tersebut
      while (name_it != this->name.end()) {
        if (*name_it == name) {
          // hapus menggunakan erase prosedur bawaan dari list
          id_it = this->id.erase(id_it);
          name_it = this->name.erase(name_it);
          category_it = this->category.erase(category_it);
          price_it = this->price.erase(price_it);
          cout << "\n✅ Data berhasil dihapus!\n";
          return;
        }
        // iterator untuk next elemen
        ++id_it;
        ++name_it;
        ++category_it;
        ++price_it;
      }
      // jika data tidak ditemukan, maka akan menampilkan pesan
      cout << "❌ Data tidak ditemukan.\n";
    }

    // mengupdate data berdasarkan nama yang akan dicari
    void updatePetShop(string name, string newName, string newCategory, int newPrice){
      // Inisialisasi iterator
      // digunakan untuk mengakses data dalam list produk
      auto name_it = this->name.begin();
      auto category_it = this->category.begin();
      auto price_it = this->price.begin();

      cout << "\n🔍 Mengupdate '" << name << "'...\n";
      // mencari dulu data nya, jika ditemukan maka update data tersebut
      while (name_it != this->name.end()) {
        if (*name_it == name) {
          // menimpa data lama dengan data baru
          *name_it = newName;
          *category_it = newCategory;
          *price_it = newPrice;
          cout << "\n✅ Data berhasil diupdate!\n";

          return;
        }
        // iterator untuk next elemen
        ++name_it;
        ++category_it;
        ++price_it;
      }
      // jika data tidak ditemukan, maka akan menampilkan pesan
      cout << "❌ Data tidak ditemukan.\n";
    }

    // destruct
    ~PetShop(){
    }
};

File: java/PetShop.java
import java.util.ArrayList; // Menggunakan ArrayList untuk menyimpan daftar hewan

// Kelas utama PetShop
public class PetShop {
    // Deklarasi ArrayList untuk menyimpan atribut hewan
    private ArrayList<Integer> id;       // Menyimpan ID hewan
    private ArrayList<String> name;      // Menyimpan nama hewan
    private ArrayList<String> category;  // Menyimpan kategori
    private ArrayList<Integer> price;    // Menyimpan harga hewan

    // Konstruktor (dijalankan saat objek dibuat)
    public PetShop() {
        // Inisialisasi list kosong
        id = new ArrayList<>();
        name = new ArrayList<>();
        category = new ArrayList<>();
        price = new ArrayList<>();

        // Menambahkan data hewan awal ke dalam list
        id.add(1);
        name.add("Kucing Persia");
        category.add("Hewan");
        price.add(98999);

        id.add(2);
        name.add("Penyu Madura");
        category.add("Hewan");
        price.add(215999);

        id.add(3);
        name.add("Katak Acumalaka");
        category.add("Hewan");
        price.add(14199);
    }

    // Metode untuk menambahkan hewan baru ke dalam daftar
    public void setPetShop(int id, String name, String category, int price) {
        this.id.add(id);
        this.name.add(name);
        this.category.add(category);
        this.price.add(price);
    }

    // Getter untuk mengambil daftar ID hewan
    public ArrayList<Integer> getId() {
        return id;
    }

    // Getter untuk mengambil daftar nama hewan
    public ArrayList<String> getName() {
        return name;
    }

    // Getter untuk mengambil daftar kategori hewan
    public ArrayList<String> getCategory() {
        return category;
    }

    // Getter untuk mengambil daftar harga hewan
    public ArrayList<Integer> getPrice() {
        return price;
    }

    // Metode untuk menampilkan daftar hewan dalam bentuk tabel yang rapi
    public void showPetShop() {
        System.out.println("\n            Daftar Hewan di Pet Shop ");
        System.out.println("--------------------------------------------------");
        // Header tabel
        System.out.printf("%-5s %-20s %-15s %-10s\n", "ID", "Name", "Category", "Price");
        System.out.println("--------------------------------------------------");

        // Loop untuk menampilkan semua hewan yang ada dalam daftar
        for (int i = 0; i < id.size(); i++) {
            System.out.printf("%-5d %-20s %-15s Rp %-10d\n", 
                id.get(i), name.get(i), category.get(i), price.get(i));
        }
        System.out.println(); // Menambahkan baris kosong agar tampilan lebih rapi
    }

    // Metode untuk mencari hewan berdasarkan nama
    public boolean findName(String searchName) {
        System.out.println("Mencari '" + searchName + "'...");

        // Loop untuk mencari nama dalam daftar
        for (int i = 0; i < name.size(); i++) {
            if (name.get(i).equals(searchName)) { // Jika nama ditemukan
                System.out.println("Data ditemukan!\n");
                System.out.printf("%-5s %-20s %-15s %-10s\n", "ID", "Name", "Category", "Price");
                System.out.println("--------------------------------------------------");
                System.out.printf("%-5d %-20s %-15s Rp %-10d\n", 
                    id.get(i), name.get(i), category.get(i), price.get(i));
                return true; // Mengembalikan true jika ditemukan
            }
        }
        System.out.println("Data tidak ditemukan.");
        return false; // Mengembalikan false jika tidak ditemukan
    }

    // Metode untuk menghapus hewan berdasarkan nama
    public void deletePetShop(String nameToDelete) {
        System.out.println("Menghapus '" + nameToDelete + "'...");
        for (int i = 0; i < name.size(); i++) {
            if (name.get(i).equals(nameToDelete)) { // Jika nama ditemukan
                id.remove(i);       // Hapus ID dari daftar
                name.remove(i);     // Hapus nama dari daftar
                category.remove(i); // Hapus kategori dari daftar
                price.remove(i);    // Hapus harga dari daftar
                System.out.println("Data berhasil dihapus!");
                return;
            }
        }
        System.out.println("Data tidak ditemukan."); // Jika tidak ditemukan
    }

    // Metode untuk memperbarui data hewan berdasarkan nama
    public void updatePetShop(String oldName, String newName, String newCategory, int newPrice) {
        System.out.println("Mengupdate '" + oldName + "'...");
        for (int i = 0; i < name.size(); i++) {
            if (name.get(i).equals(oldName)) { // Jika nama ditemukan
                name.set(i, newName);          // Update nama
                category.set(i, newCategory);  // Update kategori
                price.set(i, newPrice);        // Update harga
                System.out.println("Data berhasil diupdate!");
                return;
            }
        }
        System.out.println("Data tidak ditemukan."); // Jika tidak ditemukan
    }
}
